### ISSUE TITLE
Slack Bot integration with FunRetro

### ISSUE THREAD

**Issue by JavierLight** [funretro/distributed#231]
## Description

Fun Retro integration into Slack using a bot.

## Motivation

Would be great to be able to type /newfunretro in my Slack team and get the link for a new FunRetro (instead of creating a new one using the buttons in the webpage). A customizable Name for the Board would be passed as an argument in this command.

## Why do I need your help?

I would be more than pleased to contribute creating the Slack Bot, but I would like to know if you can find an easy way to provide this service via a new endpoint in your website. My alternative would be programatically creating a new user and appending that **AuthO user ID** to the **href.window.location** redirection, which doesn't look very elegant for me.

**Comment by glauberramos** [comment]
The problem is that we don't have server side for that app right now. So there is no way of creating an endpoint for that. But maybe we can create a simple server in order to do that. That would be great to do actually.

**Comment by JavierLight** [comment]
I'm sorry I didn't explain myself properly. What I meant is to avoid the need to:

- Press on the `New Board` button.
- Input the name of the board.
- Press again the `New Board` button.

If you could create a new endpoint being [http://funretro.github.io/distributed/newboard?name](http://funretro.github.io/distributed/newboard?name), its response being the URL of the new board, that would be perfect for me to create the Slack Bot. A simple `GET` request could be the easiest way, right?

I'm assuming this would be possible as long as there is no actual user being logged in (which currently seems to be the situation).

Thank you very much for your time :)

**Comment by glauberramos** [comment]
That's the issue, the application right now is pure static, only runs on the client side. There is no way of creating what you are asking for. There is no code running on the server side. We use a service called Firebase that handles everything, and that server is called from the client side.

We would need to create a simple node express server to handle what you are asking.

Did you understand now?

**Comment by JavierLight** [comment]
Sorry again, I misunderstood your first message. That would be awesome! Hoping that somebody can create that simple node express server (I don't even know how to start).

**Comment by glauberramos** [comment]
Yeah I think it would be nice to create this simple node express server. I will see if I can do it on my free time but right now I am quite busy. It would be nice if someone else tries it as well. Maybe @pamo or @Roshanjossey?

**Comment by juventinoromero** [comment]
Hi @glauberramos , I'm new to contributing and to this project, but I might be able to help you out with creating the node express server.

**Comment by glauberramos** [comment]
hey @juventinoromero, I think we should try first @Roshanjossey suggestion to use firebase functions. Can you install a local funretro version on your local machine and try to use firebase functions in order to create a board and return a board URL?

**Comment by juventinoromero** [comment]
@glauberramos sure, I'll give it a shot.

**Comment by juventinoromero** [comment]
Hello @glauberramos, I think I can implement firebase functions. I've taken a look to the code but can't seem to understand how the app interacts with firebase.

**Comment by Roshanjossey** [comment]
> I think I can implement firebase functions

@juventinoromero, that's awesome.

> can't seem to understand how the app interacts with firebase.

I'll share what I understood by reading code here. @glauberramos, please jump in and correct me if I'm wrong.
Currently we're only using database feature of firebase.
To know more, read through https://github.com/funretro/distributed/blob/master/js/services/firebaseService.js and https://github.com/funretro/distributed/blob/master/js/mainController.js.

`firebaseService.js` is where we've extracted out all firebase actions as utility functions. You can see usages of these functions in `mainController.js`

**Comment by juventinoromero** [comment]
thanks @Roshanjossey! I'm checking it out and it starts to make sense. I think the createNewBoard function in mainController.js is the one I need to implement in the /newRetro endpoint, isn't it?

**Comment by glauberramos** [comment]
Yeah @juventinoromero, that's the method we need to implement.

**Comment by juventinoromero** [comment]
Hey @glauberramos and @Roshanjossey , I'm not sure how to implement the createNewBoard in the firebase function. I'm not sure how angularworks and how to proceed with those $scope calls you make. Can you take a look at this file and somehow give me some guidance? https://github.com/juventinoromero/distributed/blob/newboard-endpoint/functions/index.js

**Comment by glauberramos** [comment]
Hello @juventinoromero.

I don't think you should use angular at all and any other file from the application. Your function should be separated from the rest of the application.

I suggest you to use hardcoded values. You can copy the messageTypes, use 6 for max_votes, text_editing_is_private should be true. Also you should not use firebaseService but you should call firebase direct from you file (just copy the function). Also you should not use $scope.

Does this helps you?

**Comment by juventinoromero** [comment]
@glauberramos, it does! Thanks :smile:

**Comment by juventinoromero** [comment]
@glauberramos can you have a look to some changes I've made? I'm working on figuring out why I get a request error, but I think I've made good progress. https://github.com/juventinoromero/distributed/blob/newboard-endpoint/functions/index.js

**Comment by juventinoromero** [comment]
I've fixed the error, the api successfully creates a new user and a new board. @glauberramos can you insert the correct database info in firebaseInitialization.js to check if it works?

**Comment by glauberramos** [comment]
@juventinoromero the code worked perfectly 💃, I already merged your pull request, thanks for that.
@JavierLight we now have the endpoint you need for creating the slack bot. Here is the endpoint: https://us-central1-blinding-torch-6662.cloudfunctions.net/newboard?name=test

**Comment by juventinoromero** [comment]
@glauberramos great news! 🕺 Thanks!

**Comment by JavierLight** [comment]
You guys are amazing. Thank you very much for your time and effort!!

2018-01-08 1:07 GMT+01:00 Juventino Romero <notifications@github.com>:

**Comment by Kiri23** [comment]
What is the status of this. I want to help.

**Comment by glauberramos** [comment]
I think this is on hold, if you want to implement the slack plugin please go ahead @Kiri23

**Comment by Kiri23** [comment]
But how do I do it with vanilla JavaScript because you don’t have a server/backend like for instance Node.js.

El 02/06/2018, a la(s) 9:02 a. m., Glauber Ramos <notifications@github.com<mailto:notifications@github.com>> escribió:


I think this is on hold, if you want to implement the slack plugin please go ahead @Kiri23<https://github.com/kiri23>

**Comment by glauberramos** [comment]
Firebase functions 💃

**Comment by Kiri23** [comment]
And what version of angular is this project build from. Is this angular 1 ?

El 02/06/2018, a la(s) 1:41 p. m., Glauber Ramos <notifications@github.com<mailto:notifications@github.com>> escribió:


Firebase functions 💃

**Comment by glauberramos** [comment]
You can check that on package.json
